### PR TITLE
Bug 1314864 - Update SyncView to contain new updated and new fields from the sync ping.

### PR DESCRIPTION
**Pull request by thomcc**
This updates the sync ping view to handle changes to the sync ping as well as adds support for recording new data.

This encompasses the work for [bug 1314864](https://bugzilla.mozilla.org/show_bug.cgi?id=1314864). Specific set of data changes, in no order:

1. The device constellation for a user, [from bug 1308567](https://bugzilla.mozilla.org/show_bug.cgi?id=1308567).
2. Bookmark validation data, [from bug 1267917](https://bugzilla.mozilla.org/show_bug.cgi?id=1267917) and is present in a very small number of cases. (This is currently something like 1 sync per day, for 10% of nightly users who have fewer than 100 bookmarks, although those numbers are subject to change).
3. Reporting an updated error type added in [Bug 1312320](https://bugzilla.mozilla.org/show_bug.cgi?id=1312320).
4. The ping's `application.channel` property is now recorded. No bug tracking this but it's been requested several times by e.g. adavis.
5. The server-side portion of [bug 1301389](https://bugzilla.mozilla.org/show_bug.cgi?id=1301389), which hasn't landed (I've added a patch to it so that you can see the equivalent documentation/schema changes on that bug). This is a necessary step for [bug 1289536](https://bugzilla.mozilla.org/show_bug.cgi?id=1289536).

Let me know if you have any questions.

P.S. (Apologies in advance if I messed up the copy/paste for any of the bugs listed here. Tried to be thorough, but that does lead to a large number of them)

---
This change is [Reviewable](https://reviewable.io/reviews/mozilla/telemetry-batch-view/145)

### CODE REVIEW

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by thomcc** [comment]
Whoops, fixed the PR title, sorry about that. Not sure if anything is needed to make the autolander aware.

**Comment by vitillo** [comment]
@mhammond r?

File: src/test/scala/com/mozilla/telemetry/SyncViewTest.scala [hunk 1]
```diff
@@ -164,8 +188,44 @@ class SyncViewTest extends FlatSpec with Matchers{
...
 
     secondSync.getAs[GenericRowWithSchema]("status") should be (null)
+    secondSync.getAs[GenericRowWithSchema]("devices") should be (null)
 
     val secondEngines = secondSync.getAs[mutable.WrappedArray[GenericRowWithSchema]]("engines")
     validateEngines(secondEngines, (secondPing \ "engines").extract[List[JValue]])
   }
+
+
+  // A helper to check the contents of the sync ping wiht devices
```
Comment (mhammond) [inline src/test/scala/com/mozilla/telemetry/SyncViewTest.scala]: typo: "with"

File: src/main/scala/com/mozilla/telemetry/views/SyncView.scala [hunk 2]
```diff
@@ -304,6 +322,45 @@ object SyncPingConverter {
...
+  private def validationToRow(validation: JValue): Row = validation match {
+    case JObject(_) =>
+      Row(
+        validation \ "version" match {
+          case JInt(x) => x.toLong
+          case _ => 0L
+        },
+        validation \ "checked" match {
+          case JInt(x) => x.toLong
+          case _ => 0L // Should this
```
Comment (mhammond) [inline src/main/scala/com/mozilla/telemetry/views/SyncView.scala]: this comment fragment should either be removed or completed. While checked==0 does imply something very strange, I think it makes sense to include it, simply so we can identify weirdness if it happens.

**Comment by vitillo** [comment]
lgtm

**Comment by thomcc** [comment]
So, what is needed in order to get this merged?

**Pull request merged**